src/rulesets/LinearAlgebra/norm.jl

```diff
@@ -17,55 +17,67 @@ function frule((_, Δx), ::typeof(norm), x::Number, p::Real)
     return y, ∂y
 end
 
-function rrule(
-    ::typeof(norm),
-    x::Union{StridedArray, LinearAlgebra.AbstractTriangular, Diagonal},
-    p::Real,
-)
+function rrule(::typeof(norm), x::AbstractArray{<:Number}, p::Real)
     y = LinearAlgebra.norm(x, p)
-    function norm_pullback(Δy)
-        ∂x = Thunk() do
-            return if isempty(x) || p == 0
-                zero.(x) .* (zero(y) * zero(real(Δy)))
+    function norm_pullback_p(Δy)
+        ∂x = InplaceableThunk(
+            # out-of-place versions
+            if isempty(x) || p == 0
+                @thunk(zero.(x) .* (zero(y) * zero(real(Δy))))
             elseif p == 2
-                _norm2_back(x, y, Δy)
+                @thunk(_norm2_back(x, y, Δy))
             elseif p == 1
-                _norm1_back(x, y, Δy)
+                @thunk(_norm1_back(x, y, Δy))
             elseif p == Inf
-                _normInf_back(x, y, Δy)
+                @thunk(_normInf_back(x, y, Δy))
             elseif p == -Inf
-                _normInf_back(x, y, Δy)
+                @thunk(_normInf_back(x, y, Δy))
             else
-                _normp_back_x(x, p, y, Δy)
+                @thunk(_normp_back_x(x, p, y, Δy))
+            end,
+            # in-place versions
+            if isempty(x) || p == 0
+                identity
+            elseif p == 2
+                dx -> _norm2_back!(dx, x, y, Δy)
+            elseif p == 1
+                dx -> _norm1_back!(dx, x, y, Δy)
+            elseif p == Inf
+                dx -> dx .+= _normInf_back(x, y, Δy)  # not really in-place! could perhaps be improved
+            elseif p == -Inf
+                dx -> dx .+= _normInf_back(x, y, Δy)
+            else
+                dx -> dx .+= _normp_back_x(x, p, y, Δy)
             end
-        end
+            )
         ∂p = @thunk _normp_back_p(x, p, y, Δy)
         return (NO_FIELDS, ∂x, ∂p)
     end
-    norm_pullback(::Zero) = (NO_FIELDS, Zero(), Zero())
-    return y, norm_pullback
+    norm_pullback_p(::Zero) = (NO_FIELDS, Zero(), Zero())
```

Comment:
Do we need this?
we normally don't include this because it is always the case, and we leave it to the AD to insert it (which Zygote does)

Reply:
I don't know, but this PR hasn't changed it.

Reply:
Fair enough